cirq-core/cirq/protocols/decompose_protocol_test.py

```diff
@@ -308,3 +310,40 @@ def test_decompose_tagged_operation():
         'tag',
     )
     assert cirq.decompose_once(op) == cirq.decompose_once(op.untagged)
+
+
+def test_decompose_recursive_dfs():
+    class RecursiveDecompose(cirq.Gate):
+        def __init__(self, recurse: bool = True, mock_qm: Optional[mock.Mock] = None):
+            self.recurse = recurse
+            self.mock_qm = mock.Mock() if mock_qm is None else mock_qm
```

Comment:
I see few places in Cirq where stubbing is used, lets not normalize that. Stubs can go horribly wrong => tests that are smoke tests, change detectors or worse tests that pass ragardless of the change.

lets do the test in a better way.

Reply:
In this specific case, I think its okay to use stubs since we want to test the ordering in which `cirq.decompose` will traverse the yielded `OP-TREE` and this test asserts that specifically. I don't care about the behavior of a Qubit Manager in this test -- it will become relevant when we add a qubit manager and that would be the time to add an integration test. 

I could modify the test that asserts the traversal order via some other mechanism (eg: keep a global counter and append items to a list whenever the qalloc / qfree statements are executed) but the mock is essentially doing the same thing so I don't see a reason to replace it with some other equivalent mechanism. 

While I generally appreciate the concern for not overusing mocks to write tests, I think the use here is well justified. I'll be curious to hear any alternative suggestions you have in mind for rewriting the test though.